File: frontend/src/api/models/texture.ts
import { http } from "../http";

// ==================== 类型定义 ====================

export interface TextureFile {
  id: number;
  file_type: string;
  related_id: number;
  related_type: string;
  original_url: string;
  local_path: string;
  cdn_path: string;
  file_name: string;
  file_size: number;
  width: number;
  height: number;
  format: string;
  md5: string;
  version: number;
  status: number;
  download_retry: number;
  last_error: string;
  created_at: string;
  updated_at: string;
  deleted_at?: string;
  full_url: string; // 后端自动拼接的完整 URL
}

export interface Texture {
  id: number;
  asset_id: string;
  name: string;
  description: string;
  type: number;
  authors: string;
  max_resolution: string;
  files_hash: string;
  date_published: number;
  download_count: number;
  use_count: number;
  last_used_at?: string;
  priority: number;
  sync_status: number;
  created_at: string;
  updated_at: string;
  files?: TextureFile[]; // 添加文件列表
}

export interface TextureSyncLog {
  id: number;
  sync_type: string;
  status: number;
  total_count: number;
  processed_count: number;
  success_count: number;
  fail_count: number;
  skip_count: number;
  current_asset: string;
  progress: number;
  download_speed: number;
  start_time: string;
  end_time: string;
  error_msg: string;
  created_at: string;
  updated_at: string;
}

export interface TextureListParams {
  page?: number;
  pageSize?: number; // 改为 pageSize 匹配后端
  keyword?: string;
  tag_id?: number;
  sortBy?: "use_count" | "date_published" | "created_at"; // 改为 sortBy
  order?: "asc" | "desc";
  syncStatus?: number; // 同步状态: 0=未同步 1=同步中 2=已同步 3=失败
}

export interface TextureListResponse {
  list: Texture[]; // 改为 list 匹配后端返回
  total: number;
  page: number;
  pageSize: number; // 改为 pageSize
}

export interface SyncProgress {
  log_id: number;
  status: number;
  progress: number;
  current_asset: string;
  processed_count: number;
  total_count: number;
  download_speed: number;
}

// ==================== API 方法 ====================

/**
 * 获取材质列表
 */
export const getTextureList = (params?: TextureListParams) => {
  return http.get<TextureListResponse>("textures", { params });
};

/**
 * 获取材质详情
 */
export const getTextureDetail = (assetId: string) => {
  return http.get<Texture>(`textures/${assetId}`);
};

/**
 * 记录材质使用
 */
export const recordTextureUse = (assetId: string) => {
  return http.post(`textures/${assetId}/use`);
};

/**
 * 触发同步
 */
export const triggerSync = (data?: { sync_type?: string }) => {
  return http.post<{ log_id: number }>("textures/sync", data);
};

/**
 * 获取同步进度
 */
export const getSyncProgress = () => {
  return http.get<SyncProgress>("textures/sync/progress");
};

/**
 * 获取同步状态
 */
export const getSyncStatus = (logId: number) => {
  return http.get<TextureSyncLog>(`textures/sync/status/${logId}`);
};

/**
 * 获取同步日志列表
 */
export const getSyncLogs = (params?: { page?: number; pageSize?: number }) => {
  return http.get<{ items: TextureSyncLog[]; total: number }>(
    "textures/sync/logs",
    { params },
  );
};

File: models/texture.go
package models

import "time"

// Texture 材质表
type Texture struct {
	ID                uint       `gorm:"primaryKey" json:"id"`
	AssetID           string     `gorm:"uniqueIndex;size:100" json:"asset_id"`
	Name              string     `gorm:"size:200;index" json:"name"`
	Description       string     `gorm:"type:text" json:"description"`
	Type              int        `json:"type"`
	Authors           string     `gorm:"type:text" json:"authors"`
	MaxResolution     string     `gorm:"size:50" json:"max_resolution"`
	FilesHash         string     `gorm:"size:100;index" json:"files_hash"`
	DatePublished     int64      `gorm:"index" json:"date_published"`
	DownloadCount     int        `json:"download_count"`
	UseCount          int        `gorm:"default:0;index" json:"use_count"`
	LastUsedAt        *time.Time `json:"last_used_at"`
	Priority          int        `gorm:"default:0;index" json:"priority"`
	SyncStatus        int        `gorm:"default:0;index" json:"sync_status"`
	DownloadCompleted bool       `gorm:"default:false;index" json:"download_completed"` // 是否已完成下载
	Source            string     `gorm:"size:20;index;default:'polyhaven'" json:"source"` // 数据来源: polyhaven, ambientcg
	TextureTypes      string     `gorm:"type:text" json:"texture_types"`                // 包含的贴图类型，逗号分隔，如: "Diffuse,Rough,Normal"
	CreatedAt         time.Time  `json:"created_at"`
	UpdatedAt         time.Time  `json:"updated_at"`
}

// TextureFile 材质文件关联表
type TextureFile struct {
	ID         uint      `gorm:"primaryKey" json:"id"`
	TextureID  uint      `gorm:"index" json:"texture_id"`
	FileID     uint      `gorm:"index" json:"file_id"`
	MapType    string    `gorm:"size:50;index" json:"map_type"`
	Resolution string    `gorm:"size:20" json:"resolution"`
	CreatedAt  time.Time `json:"created_at"`
}

// TextureSyncLog 同步记录表
type TextureSyncLog struct {
	ID             uint      `gorm:"primaryKey" json:"id"`
	SyncType       string    `gorm:"size:20" json:"sync_type"`
	Status         int       `gorm:"default:0;index" json:"status"`
	TotalCount     int       `json:"total_count"`
	ProcessedCount int       `json:"processed_count"`
	SuccessCount   int       `json:"success_count"`
	FailCount      int       `json:"fail_count"`
	SkipCount      int       `json:"skip_count"`
	CurrentAsset   string    `gorm:"size:100" json:"current_asset"`
	Progress       float64   `json:"progress"`
	DownloadSpeed  float64   `json:"download_speed"`
	StartTime      time.Time `json:"start_time"`
	EndTime        time.Time `json:"end_time"`
	ErrorMsg       string    `gorm:"type:text" json:"error_msg"`
	CreatedAt      time.Time `json:"created_at"`
	UpdatedAt      time.Time `json:"updated_at"`
}

// DownloadQueue 下载队列表
type DownloadQueue struct {
	ID          uint       `gorm:"primaryKey" json:"id"`
	FileID      uint       `gorm:"index" json:"file_id"`
	TextureID   uint       `gorm:"index" json:"texture_id"`
	Priority    int        `gorm:"default:5;index" json:"priority"`
	Status      int        `gorm:"default:0;index" json:"status"`
	RetryCount  int        `gorm:"default:0" json:"retry_count"`
	MaxRetry    int        `gorm:"default:3" json:"max_retry"`
	ErrorMsg    string     `gorm:"type:text" json:"error_msg"`
	ScheduledAt time.Time  `gorm:"index" json:"scheduled_at"`
	StartedAt   *time.Time `json:"started_at"`
	CompletedAt *time.Time `json:"completed_at"`
	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
}

// TextureMetrics 系统指标表
type TextureMetrics struct {
	ID              uint      `gorm:"primaryKey" json:"id"`
	Date            string    `gorm:"uniqueIndex;size:10" json:"date"`
	TotalTextures   int       `json:"total_textures"`
	TotalFiles      int       `json:"total_files"`
	TotalSize       int64     `json:"total_size"`
	DownloadCount   int       `json:"download_count"`
	DownloadSize    int64     `json:"download_size"`
	FailedCount     int       `json:"failed_count"`
	AvgDownloadTime float64   `json:"avg_download_time"`
	CreatedAt       time.Time `json:"created_at"`
	UpdatedAt       time.Time `json:"updated_at"`
}
